diff --git a/src/plugins/vmxnet3/vmxnet3.c b/src/plugins/vmxnet3/vmxnet3.c
--- a/src/plugins/vmxnet3/vmxnet3.c
+++ b/src/plugins/vmxnet3/vmxnet3.c
@@ -180,6 +180,7 @@ vmxnet3_write_mac (vmxnet3_device_t * vd)
 static clib_error_t *
 vmxnet3_provision_driver_shared (vlib_main_t * vm, vmxnet3_device_t * vd)
 {
+  clib_error_t *error;
   vmxnet3_shared *shared;
   u64 shared_dma;
   u16 qid, rid;
@@ -192,6 +193,9 @@ vmxnet3_provision_driver_shared (vlib_main_t * vm, vmxnet3_device_t * vd)
   if (vd->driver_shared == 0)
     return vlib_physmem_last_error (vm);
 
+  if ((error = vlib_pci_map_dma (vm, vd->pci_dev_handle, vd->driver_shared)))
+    return error;
+
   clib_memset (vd->driver_shared, 0, sizeof (*vd->driver_shared));
 
   vec_foreach_index (qid, vd->txqs)
@@ -286,6 +290,7 @@ vmxnet3_disable_interrupt (vmxnet3_device_t * vd)
 static clib_error_t *
 vmxnet3_rxq_init (vlib_main_t * vm, vmxnet3_device_t * vd, u16 qid, u16 qsz)
 {
+  clib_error_t *error;
   vmxnet3_rxq_t *rxq;
   vmxnet3_rx_stats *rxs;
   u16 rid;
@@ -307,6 +312,10 @@ vmxnet3_rxq_init (vlib_main_t * vm, vmxnet3_device_t * vd, u16 qid, u16 qsz)
       if (rxq->rx_desc[rid] == 0)
 	return vlib_physmem_last_error (vm);
 
+      if ((error = vlib_pci_map_dma (vm, vd->pci_dev_handle,
+				     rxq->rx_desc[rid])))
+	return error;
+
       clib_memset (rxq->rx_desc[rid], 0, qsz * sizeof (*rxq->rx_desc[rid]));
     }
   rxq->rx_comp =
@@ -315,6 +324,9 @@ vmxnet3_rxq_init (vlib_main_t * vm, vmxnet3_device_t * vd, u16 qid, u16 qsz)
   if (rxq->rx_comp == 0)
     return vlib_physmem_last_error (vm);
 
+  if ((error = vlib_pci_map_dma (vm, vd->pci_dev_handle, rxq->rx_comp)))
+    return error;
+
   clib_memset (rxq->rx_comp, 0, qsz * sizeof (*rxq->rx_comp));
   for (rid = 0; rid < VMXNET3_RX_RING_SIZE; rid++)
     {
@@ -333,6 +345,7 @@ vmxnet3_rxq_init (vlib_main_t * vm, vmxnet3_device_t * vd, u16 qid, u16 qsz)
 static clib_error_t *
 vmxnet3_txq_init (vlib_main_t * vm, vmxnet3_device_t * vd, u16 qid, u16 qsz)
 {
+  clib_error_t *error;
   vmxnet3_txq_t *txq;
   vmxnet3_tx_stats *txs;
   u32 size;
@@ -355,6 +368,9 @@ vmxnet3_txq_init (vlib_main_t * vm, vmxnet3_device_t * vd, u16 qid, u16 qsz)
   if (txq->tx_desc == 0)
     return vlib_physmem_last_error (vm);
 
+  if ((error = vlib_pci_map_dma (vm, vd->pci_dev_handle, txq->tx_desc)))
+    return error;
+
   clib_memset (txq->tx_desc, 0, size);
 
   size = qsz * sizeof (*txq->tx_comp);
@@ -363,6 +379,9 @@ vmxnet3_txq_init (vlib_main_t * vm, vmxnet3_device_t * vd, u16 qid, u16 qsz)
   if (txq->tx_comp == 0)
     return vlib_physmem_last_error (vm);
 
+  if ((error = vlib_pci_map_dma (vm, vd->pci_dev_handle, txq->tx_comp)))
+    return error;
+
   clib_memset (txq->tx_comp, 0, size);
   vec_validate_aligned (txq->tx_ring.bufs, txq->size, CLIB_CACHE_LINE_BYTES);
   txq->tx_ring.gen = VMXNET3_TXF_GEN;
@@ -382,6 +401,7 @@ static const u8 vmxnet3_rss_key[VMXNET3_RSS_MAX_KEY_SZ] = {
 static clib_error_t *
 vmxnet3_rss_init (vlib_main_t * vm, vmxnet3_device_t * vd)
 {
+  clib_error_t *error;
   vmxnet3_rss_shared *rss;
   size_t size = sizeof (*rss);
   u8 i;
@@ -390,6 +410,9 @@ vmxnet3_rss_init (vlib_main_t * vm, vmxnet3_device_t * vd)
   if (vd->rss == 0)
     return vlib_physmem_last_error (vm);
 
+  if ((error = vlib_pci_map_dma (vm, vd->pci_dev_handle, vd->rss)))
+    return error;
+
   clib_memset (vd->rss, 0, size);
   rss = vd->rss;
   rss->hash_type =
@@ -493,6 +516,9 @@ vmxnet3_device_init (vlib_main_t * vm, vmxnet3_device_t * vd,
   if (vd->queues == 0)
     return vlib_physmem_last_error (vm);
 
+  if ((error = vlib_pci_map_dma (vm, vd->pci_dev_handle, vd->queues)))
+    return error;
+
   clib_memset (vd->queues, 0, size);
 
   if (vd->num_rx_queues > 1)
@@ -733,6 +759,8 @@ vmxnet3_create_if (vlib_main_t * vm, vmxnet3_create_if_args_t * args)
     vd->flags |= VMXNET3_DEVICE_F_ELOG;
 
   vd->pci_dev_handle = h;
+  if (vlib_pci_supports_virtual_addr_dma (vm, h))
+    vd->flags |= VMXNET3_DEVICE_F_IOVA;
   vd->numa_node = vlib_pci_get_numa_node (vm, h);
   vd->num_intrs = vd->num_rx_queues + 1;	// +1 for the event interrupt

